samza-aws/src/main/java/org/apache/samza/system/kinesis/descriptors/KinesisInputDescriptor.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.samza.serializers.NoOpSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.system.descriptors.InputDescriptor;
+import org.apache.samza.system.descriptors.SystemDescriptor;
+import org.apache.samza.system.kinesis.KinesisConfig;
+
+
+/**
+ * A {@link KinesisInputDescriptor} can be used for specifying Samza and Kinesis specific properties of Kinesis
+ * input streams.
+ * <p>
+ * Use {@link KinesisSystemDescriptor#getInputDescriptor} to obtain an instance of this descriptor.
+ * <p>
+ * Stream properties provided in configuration override corresponding properties specified using a descriptor.
+ *
+ * @param <StreamMessageType> type of messages in this stream
+ */
+public class KinesisInputDescriptor<StreamMessageType>
+    extends InputDescriptor<StreamMessageType, KinesisInputDescriptor<StreamMessageType>> {
+  private Optional<String> accessKey = Optional.empty();
+  private Optional<String> secretKey = Optional.empty();
+  private Optional<String> region = Optional.empty();
+  private Map<String, String> kclConfig = Collections.emptyMap();
+
+
+  /**
+   * Constructs an {@link InputDescriptor} instance.
+   *
+   * @param streamId id of the stream
+   * @param valueSerde serde the values in the messages in the stream
+   * @param systemDescriptor system descriptor this stream descriptor was obtained from
+   */
+  KinesisInputDescriptor(String streamId, Serde valueSerde, SystemDescriptor systemDescriptor) {
+    super(streamId, KVSerde.of(new NoOpSerde<>(), valueSerde), systemDescriptor, null);
+  }
+
+  /**
+   * Kinesis region for the system stream.
+   * @param region Kinesis region
+   * @return this input descriptor
+   */
+  public KinesisInputDescriptor<StreamMessageType> withRegion(String region) {
+    this.region = Optional.of(region);
+    return this;
+  }
+
+  /**
+   * Kinesis access key name for the system stream.
+   * @param accessKey Kinesis access key name
+   * @return this input descriptor
+   */
+  public KinesisInputDescriptor<StreamMessageType> withAccessKey(String accessKey) {
+    this.accessKey = Optional.of(accessKey);
+    return this;
+  }
+
+  /**
+   * Kinesis secret key name for the system stream.
+   * @param secretKey Kinesis secret key
+   * @return this input descriptor
+   */
+  public KinesisInputDescriptor<StreamMessageType> withSecretKey(String secretKey) {
+    this.secretKey = Optional.of(secretKey);
+    return this;
+  }
+
+  /**
+   * KCL (Kinesis Client Library) config for the system stream. This is not required by default.
+   * @param kclConfig A map of specified KCL configs
+   * @return this input descriptor
+   */
+  public KinesisInputDescriptor<StreamMessageType> withKCLConfig(Map<String, String> kclConfig) {
+    this.kclConfig = kclConfig;
+    return this;
+  }
+
+  @Override
+  public Map<String, String> toConfig() {
+    final Map<String, String> config = new HashMap<>(super.toConfig());
+
+    final String systemName = getSystemName();
```

Comment:
nit: The rest of the code-base does not use `final` for local variables since it's a bit boiler-platey.

Reply:
@vjagadish1989 : using final is good habit and should be encouraged in samza code base. I don't think your suggestion makes sense here.

Reply:
Style and good practice discussion will never end. We should talk more about these in team meeting and agree upon things together.  I will just go for consistency for now